### PR TITLE
Update the test framework

### LINKED ISSUE #5: The return value of "from_path" doesn't match the comments in the function
From rms-webtools created by [juzen2003](https://github.com/juzen2003): SETI/rms-webtools#24  

- Current from_path expected results based on comments:
  - 'COISS_2001.targz' --> 'archives-volumes/COISS_2xxx/COISS_2001.tar.gz'
  - 'COISS_2001_previews.targz' --> 'archives-previews/COISS_2xxx/COISS_2001_previews.tar.gz'
  - 'COISS_0xxx_tar.gz' -->  'archives-volumes/COISS_2xxx'

- Actual return results from function:
  - 'COISS_2001.targz' --> previews/COISS_2xxx/COISS_2001
  - 'COISS_2001_previews.targz' --> volumes/COISS_2xxx/COISS_2001
  - 'COISS_0xxx_tar.gz' --> 'volumes/COISS_0xxx'

### LINKED ISSUE #7: Code coverage shows potential bugs/missing items in tests
From rms-webtools created by [rfrenchseti](https://github.com/rfrenchseti): SETI/rms-webtools#52  

- rules/COCIRS_xxx.py never executes the loop at 1108 or the if at 1116.
- rules/COUVIS_0xxx.py doesn't exercise various branches in DATA_SET_ID()
- rules/COVIMS_0xxx.py doesn't exercise various branches in OPUS_ID_TO_PRIMARY_LOGICAL_PATH()
- filename_keylen is never tested
- tests/test_pdsfile_blackbox.py The clause at line 1271 is never executed
- tests/test_pdsfile_blackbox.py The loop at line 3042 never starts
- tests/test_pdsfile_whitebox.py The loop at 877 never starts

### LINKED ISSUE #40: PdsFile needs a different test framework
Right now adding new rules to PdsFile is a bit of a nightmare, because adding new unit tests involves:

- Starting a Python session
- Loading the holdings by hand
- Creating PdsFile objects for each test case by hand
- Calling the appropriate PdsFile access function (like `opus_products`) by hand
- Copy/pasting the result into the PdsFile source code by hand

I suggest that we change the PdsFile testing framework to be more like that used by OPUS, where each expected result is stored in a different file in a `test_results` directory. For example, the `opus_products` tests for `COUVIS_8001` could be stored in files like `test_results/COVUS_8001/opus_products/<primary filespec>`. The current `@parametrize` tests would have their expected results removed and in their place put the filename where the expect results are kept. Then there would be a mode/global variable where instead of comparing the answers, the answers retrieved from PdsFile are written to the `test_results` directory to create the golden copies in the first place.

I strongly suggest we do this before going to the effort of adding rules for new data products.

### LINKED ISSUE #41: Create a command line utility to make it easier to write rules
Currently when adding rules to `PdsFile`, one must either `import pdsfile` in iPython, manually preload the cache, instantiate `PdsFile` objects, and then look at their properties (like `opus_products`) [and then exit iPython and repeat when updating the rules], or run OPUS import and then use the OPUS UI to look at the products for particular products [and then repeat when updating the rules]. Both methods are slow, painful, and error-prone.

I suggest the creation of a small command-line utility (or series of utilities) such as:

```sh
python show_opus_products.py <abspath>
```

which would then do all the necessary steps to find the `opus_products` and then display them in a pretty tabular way (not just dump a dictionary). Similar tools could be written for other `PdsFile` attributes.

This would make it *much* easier to write new rules and verify their correctness.

### LINKED ISSUE #42: Create top-level versions of some PdsFile functions
Some of the old `PdsFile` functions like `set_logger` have moved to the `Pds3File` and `Pds4File` subclasses, and no longer exist in the superclass. However, you rarely want to set a different logger for the two subclasses, so it would be nice to have a method in the superclass that sets the logger for both. Similar methods would be useful for `use_shelves_only` and `require_shelves` and maybe others.

### PR DESCRIPTION
**Pull request by juzen2003**
Update the test framework so that it's easier to write and update tests. 
- Fixed #40
  - Modfiy the test framework to either compare results with golden copies or update the golden copies with the latest output (currently opus products and associated abspaths has been updated to use the new test framework). Here are the example commands:
    - Update all golden copies using '--update' param
        ```
        pytest pdsfile/pds3file/rules/* --update
        pytest pdsfile/pds4file/rules/* --update
        ```
        or
        `./run_tests_coverage.sh --update`
    - Update all golden copies of VG_28xx
        `pytest pdsfile/pds3file/rules/VG_28xx.py --update`
    - Update golden copy of VG_28xx opus products test  
        `pytest pdsfile/pds3file/rules/VG_28xx.py::test_opus_products --update`
- Fixed #41
  - Create a standalone script to display more readable opus products output. By default it will display the opus products output in table format. Here are example commands:
    - Given multiple paths, could be abspath or logical path in either pds3 or pds4 (`--paths`), display a table with opus type in the first column and its corresponding pdsfile list in the second column:
    `python3 utility/show_opus_products.py --path '/Users/yu-jenchang/Dropbox (SETI Institute)/Shared-OPUS/pdsdata/holdings/volumes/COCIRS_5xxx/COCIRS_5408/DATA/POIDATA/POI0408010000_FP1.LBL' volumes/COISS_1xxx/COISS_1001/data/1294561143_1295221348/W1294561202_1.IMG`
    or
    `python3 utility/show_opus_products.py --paths '/Users/yu-jenchang/Dropbox (SETI Institute)/Shared-OPUS/pdsdata/holdings/volumes/COCIRS_5xxx/COCIRS_5408/DATA/POIDATA/POI0408010000_FP1.LBL' bundles/uranus_occs_earthbased/uranus_occ_u103_eso_220cm/data/rings/u103_eso_220cm_2200nm_radius_beta_ingress_100m.xml`
    - Given a path (`--paths`), display output in pprint format (`--pprint`), the result of pprint format is handy to diff with the golden copies::
    `python3 utility/show_opus_products.py --paths volumes/COCIRS_5xxx/COCIRS_5408/DATA/POIDATA/POI0408010000_FP1.LBL --pprint`
    - Display output in tradition dictionary format (`--raw`):
    `python3 utility/show_opus_products.py --paths volumes/COCIRS_5xxx/COCIRS_5408/DATA/POIDATA/POI0408010000_FP1.LBL --raw`
    - Display output in a narrow table with opus type in top row and its corresponding pdsfile list in the second row (this is for files with long names):
    `python3 utility/show_opus_products.py --paths volumes/COCIRS_5xxx/COCIRS_5408/DATA/POIDATA/POI0408010000_FP1.LBL volumes/COISS_1xxx/COISS_1001/data/1294561143_1295221348/W1294561202_1.IMG`
    - Display output of the given opus types in table format (``--opus-types`). It will filter out the wrong opus type (xxx in the example) and display the correct ones only:
    `python3 utility/show_opus_products.py --paths volumes/COCIRS_5xxx/COCIRS_5408/DATA/POIDATA/POI0408010000_FP1.LBL --opus-types browse_thumb cocirs_browse_saturn cocirs_ispm xxx`
    - If all the given opus types are wrong, it will print a valid list of opus types and exit the program:
    `python3 utility/show_opus_products.py --paths volumes/COCIRS_5xxx/COCIRS_5408/DATA/POIDATA/POI0408010000_FP1.LBL --opus-types wrong1 wrong2`
- Fixed #42
  - Create these top level functions in PdsFile class to set variables in Pds3File & Pds4File subclasses
  ```
  use_shelves_only
  require_shelves
  set_logger
  set_easylogger
  ```
- Fixed #7
  - Make sure test suite coverage is 100% by doing these:
      - update .coveragerc by removing omit */tests/*, this will show us the test suite coverage.
      - Add blackbox/whitebox tests 
      - Fixed bugs in the test suite
      - Remove functions not used
      - Remove the code branch that will never be reach.
- Fixed #5
    - For the case 'checksums/archives/whatever':
      - In `_complete`, make sure 'checksums-archives-.*' category is properly cached, please see the comment in `_complete` function (line 3446-3463 in `pdsfile.py`)
    - For the case that ends with '.targz', like 'COISS_2001.targz':
      - In `from_path`, if bundlename is not the key to the CACHE, we try to figure out the bundleset based on the basename, and search the CACHE again using the bundleset as the key. (line 4214-4246 in `pdsfile.py`)
    - For the case with /version, like 'COISS_0xxx/v1':
      - In `from_path`, use regex to detector the slash version string and replace it with underscore version (line 4032-4037 in `pdsfile.py`)
- Status of other bugs:
    - #37: Clean up tests in the rules. Now each volset will run the same opus products and associated abspaths tests ( under `pdsfile/pdsfile_test_helper.py`).   `test_opus_id_to_primary_logical_path` (mostly added by Mark) is run in each volset as well, but the function is specific to each volset. There is `test_opus_products_count` function run in three volset only (added by Mark), not sure why we need this since we already have opus products tests. Maybe we should remove them? 
    - #6: Some old pickle files like ones created in 2020 doesn't have sorted keys, for example: 'holdings/_infoshelf-archives-volumes/COISS_3xxx_info.pickle'
    For ones created in 2021, they will have sorted keys, for example: 'holdings/_infoshelf-archives-volumes/COISS_0xxx_info.pickle' and 'holdings/_infoshelf-archives-volumes/HSTJx_xxxx_info.pickle', we can remove the sorting at line 5085 of `_get_shelf` function in `pdsfile.py` if all the shelf files are generated after 2021.
    Example code block to read the keys of pickle files: 
  ```
  import pickle
  shelf_path = '/Users/yu-jenchang/Dropbox (SETI Institute)/Shared-OPUS/pdsdata/holdings/_infoshelf-archives-volumes/HSTJx_xxxx_info.pickle'
  with open(shelf_path, 'rb') as f:
          shelf = pickle.load(f)
  print(shelf.keys())
  ```

### CODE REVIEW

**Comment by juzen2003** [comment]
8/29/24 update: (Top comment is updated)
- Change all golden copies from `.py` to `.txt` and store contents using `pprint` so that they can be easily diff to compare results.
- Add and update more options for `show_opus_products.py`, example commands are added in the top comments (issue 41 area).

File: COCIRS_xxxx/associated_abspaths/volumes_POI0408010000_FP1.py [hunk 1]
```diff
@@ -0,0 +1 @@
...volumes/COCIRS_5xxx/COCIRS_5408/DATA/APODSPEC/SPEC0408010000_FP1.LBL', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/ISPMDATA/ISPM0408010000_FP1.TAB', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/ISPMDATA/ISPM0408010000_FP1.LBL', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/TARDATA/TAR0408010000_FP1.TAB', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/TARDATA/TAR0408010000_FP1.LBL', 'volumes/COCIRS_5xxx/COCIRS_5408/BROWSE/TARGETS/IMG0408010000_FP1.PNG', 'volumes/COCIRS_5xxx/COCIRS_5408/BROWSE/TARGETS/IMG0408010000_FP1.LBL', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/POIDATA/POI0408010000_FP1.TAB', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/POIDATA/POI0408010000_FP1.LBL', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/GEODATA/GEO0408010000_611.LBL', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/GEODATA/GEO0408010000_611.TAB', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/GEODATA/GEO0408010000_617.LBL', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/GEODATA/GEO0408010000_617.TAB', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/GEODATA/GEO0408010000_699.LBL', 'volumes/COCIRS_5xxx/COCIRS_5408/DATA/GEODATA/GEO0408010000_699.TAB', 'volumes/COCIRS_5xxx/COCIRS_5408/BROWSE/SATURN/POI0408010000_FP1.LBL', 'volumes/COCIRS_5xxx/COCIRS_5408/BROWSE/SATURN/POI0408010000_FP1.PNG']
```
Comment (rfrenchseti) [inline COCIRS_xxxx/associated_abspaths/volumes_POI0408010000_FP1.py:1]: Shouldn't this be pretty-printed?

File: COCIRS_xxxx/opus_products/000IA_PRESOI001____RI____699_F4_038P.py [hunk 2]
```diff
@@ -0,0 +1 @@
...IVE/000IA_PRESOI001____RI____699_F4_038P_small.jpg'], ('browse', 10, 'browse_thumb', 'Browse Image (thumbnail)', False): ['previews/COCIRS_0xxx/COCIRS_0406/DATA/CUBE/POINT_PERSPECTIVE/000IA_PRESOI001____RI____699_F4_038P_thumb.jpg'], ('metadata', 4, 'cube_index', 'Cube Index', False): ['metadata/COCIRS_0xxx/COCIRS_0406/COCIRS_0406_cube_point_index.tab', 'metadata/COCIRS_0xxx/COCIRS_0406/COCIRS_0406_cube_point_index.lbl'], ('metadata', 8, 'supplemental_index', 'Supplemental Index', False): ['metadata/COCIRS_0xxx/COCIRS_0406/COCIRS_0406_cube_point_supplemental_index.tab', 'metadata/COCIRS_0xxx/COCIRS_0406/COCIRS_0406_cube_point_supplemental_index.lbl'], ('Cassini CIRS', 700, 'cocirs_documentation', 'Documentation', False): ['documents/COCIRS_0xxx/Volume-SIS.pdf', 'documents/COCIRS_0xxx/Spectral-Cube-SIS.pdf', 'documents/COCIRS_0xxx/Flasar-etal-2004-SSR.pdf', 'documents/COCIRS_0xxx/FOV-Overview.pdf', 'documents/COCIRS_0xxx/Data-Product-SIS.pdf', 'documents/COCIRS_0xxx/Chan-etal-2015.pdf', 'documents/COCIRS_0xxx/Cassini-CIRS-Final-Report.pdf', 'documents/COCIRS_0xxx/Calibration-Equations.pdf', 'documents/COCIRS_0xxx/CIRS-Users-Guide.pdf', 'documents/COCIRS_0xxx/CIRS-Interference.pdf']}
```
Comment (rfrenchseti) [inline COCIRS_xxxx/opus_products/000IA_PRESOI001____RI____699_F4_038P.py:1]: Shouldn't this be pretty-printed?

File: pdsfile/pds4file/rules/uranus_occs_earthbased.py [hunk 3]
```diff
@@ -265,7 +265,7 @@
          r'bundles/\1/uranus_occ_support/document/user_guide/*-user-guide.pdf',
          r'bundles/\1/uranus_occ_support/document/user_guide/*-user-guide.xml',
          r'bundles/\1/uranus_occ_support/document/user_guide/*.pro',
-         r'bundles/\1/uranus_occ_support/document/user_guide/*.py',
+         r'bundles/\1/uranus_occ_support/document/user_guide/*.txt',
```
Comment (rfrenchseti) [inline pdsfile/pds4file/rules/uranus_occs_earthbased.py:268]: There are no `*.txt` files there. `*.py` was correct. The Python programs are part of the documentation.

File: pdsfile/pds4file/rules/uranus_occs_earthbased.py [hunk 4]
```diff
@@ -491,459 +491,29 @@ class uranus_occs_earthbased(pds4file.Pds4File):
...
-    opus_products_test(input_path, expected)
+def test_opus_products(request, input_path, expected):
+    update = request.config.option.update
+    opus_products_test(pds4file.Pds4File, input_path, TEST_RESULTS_DIR+expected, update)
 
 @pytest.mark.parametrize(
     'input_path,category,expected',
     [
-        ('uranus_occs_earthbased/uranus_occ_u0_kao_91cm/data/atmosphere/u0_kao_91cm_734nm_counts-v-time_atmos_egress.xml',
+        ('bundles/uranus_occs_earthbased/uranus_occ_u0_kao_91cm/data/atmosphere/ u0_kao_91cm_734nm_counts-v-time_atmos_egress.xml',
```
Comment (rfrenchseti) [inline pdsfile/pds4file/rules/uranus_occs_earthbased.py:511]: Is there supposed to be a space in the filename?

File: pdsfile/pdsfile.py [hunk 5]
```diff
@@ -620,6 +620,72 @@ def new_pdsfile(self, key=None, copypath=False):
 
         return this
 
+    ######################################################################################
+    # Set parameters for both Pds3File and Pds4File
+    ######################################################################################
+    @classmethod
+    def use_shelves_only(cls, pds3_status=True, pds4_status=True):
```
Comment (rfrenchseti) [inline pdsfile/pdsfile.py:627]: For this and `require_shelves` the point is to set both subclasses to the same value. If there are two parameters, we might as well just call the two subclass functions directly.

File: pdsfile/pdsfile.py [hunk 6]
```diff
@@ -620,6 +620,72 @@ def new_pdsfile(self, key=None, copypath=False):
...
+    def set_easylogger(cls):
+        """Log all messages directly to stdout.
+
+        Keyword arguments:
+            cls -- the class calling the other methods inside the function
+        """
+
+        subclasses = cls.__subclasses__()
+        for child_class in subclasses:
+             child_class.set_logger(pdslogger.EasyLogger())
```
Comment (rfrenchseti) [inline pdsfile/pdsfile.py:687]: These should probably be set to the set EasyLogger.

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
- The automated tests aren't running because the branch isn't up-to-date with `main`
- When I run the tests myself against either the OPUS Dropbox or the real holdings I get a bunch of failures.

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
For `show_opus_products`:
- Perhaps we don't need the separate abspath and logical path command line options? Just try one and then the other and use whichever doesn't throw an error.
- At the same time, it would be nice to be able to specify more than one path on the command line and just have it cycle through them printing the opus products for each.
- Probably no point in printing the big "OPUS PRODUCTS OUTPUT" header since it's obvious they are opus products.

**Comment by juzen2003** [comment]
9/4/24 update: (Top comment is updated)
- Update based on github comments
- Fixed issue 5 and 7
- Update the status of issue 6

**Review by rfrenchseti — APPROVED** [review]

**Pull request merged**